### ISSUE TITLE
Crashes with cluster-autoscaler 1.0.3 and AWS during cluster start-up

### ISSUE THREAD

**Issue by whereisaaron** [kubernetes/autoscaler#518]
During cluster start-up the `cluster-autoscaler` may have difficulty auto-discovering AWS ASGs. In these cases, rather than retry, the `cluster-autoscaler` can crash.

I'm launching an cluster-autoscaler pod running on controller/master nodes early during cluster creation and I noticed it crashes a couple of times before settling down. Checking the logs it starts up fine and establishes all its cluster watches. Then a couple minutes after start-up to goes to query AWS to discover ASGs and experiences an TCP timeout. Given this is during the start-up the worker ASGs are still being created, there may be none at this point. The timeout triggers and outright crash.

```
I1228 18:49:16.303288       1 auto_scaling.go:138] Failed to describe ASG tags for keys [k8s.io/cluster-autoscaler/enabled kubernetes.io/cluster/gnat] : RequestError: send request failed
caused by: Post https://autoscaling.us-east-2.amazonaws.com/: dial tcp: i/o timeout
F1228 18:49:16.303338       1 cloud_provider_builder.go:111] Failed to create AWS cloud provider: Failed to get ASGs: RequestError: send request failed
caused by: Post https://autoscaling.us-east-2.amazonaws.com/: dial tcp: i/o timeout
goroutine 77 [running]:
>crash<
...
k8s.io/autoscaler/cluster-autoscaler/cloudprovider/builder.CloudProviderBuilder.Build(0x7fff352eb876, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/builder/cloud_provider_builder.go:111 +0x756
```

Larger selection of log:

```
...
I1228 18:47:01.215194       1 leaderelection.go:184] successfully acquired lease kube-system/cluster-autoscaler
I1228 18:47:01.300081       1 factory.go:33] Event(v1.ObjectReference{Kind:"Endpoints", Namespace:"kube-system", Name:"cluster-autoscaler", UID:"3323df5f-ebff-11e7-bb4b-023c3416291e", APIVersion:"v1", ResourceVersion:"920", FieldPath:""}): type: 'Normal' reason: 'LeaderElection' cluster-autoscaler-74f958855-kthlz became leader
I1228 18:47:01.300952       1 predicates.go:123] Using predicate PodFitsResources
I1228 18:47:01.300967       1 predicates.go:123] Using predicate GeneralPredicates
I1228 18:47:01.300972       1 predicates.go:123] Using predicate PodToleratesNodeTaints
I1228 18:47:01.300976       1 predicates.go:123] Using predicate MaxGCEPDVolumeCount
I1228 18:47:01.300980       1 predicates.go:123] Using predicate CheckNodeCondition
I1228 18:47:01.300983       1 predicates.go:123] Using predicate CheckNodeMemoryPressure
I1228 18:47:01.300987       1 predicates.go:123] Using predicate MaxEBSVolumeCount
I1228 18:47:01.300991       1 predicates.go:123] Using predicate MaxAzureDiskVolumeCount
I1228 18:47:01.300996       1 predicates.go:123] Using predicate NoDiskConflict
I1228 18:47:01.301000       1 predicates.go:123] Using predicate CheckNodeDiskPressure
I1228 18:47:01.301005       1 predicates.go:123] Using predicate MatchInterPodAffinity
I1228 18:47:01.301009       1 predicates.go:123] Using predicate NoVolumeZoneConflict
I1228 18:47:01.301013       1 predicates.go:123] Using predicate NoVolumeNodeConflict
I1228 18:47:01.301017       1 predicates.go:123] Using predicate ready
I1228 18:47:01.301213       1 reflector.go:202] Starting reflector *v1beta1.DaemonSet (1h0m0s) from k8s.io/autoscaler/cluster-autoscaler/utils/kubernetes/listers.go:293
I1228 18:47:01.301243       1 reflector.go:240] Listing and watching *v1beta1.DaemonSet from k8s.io/autoscaler/cluster-autoscaler/utils/kubernetes/listers.go:293
I1228 18:47:01.301398       1 reflector.go:202] Starting reflector *v1.ReplicationController (0s) from k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/client-go/informers/factory.go:73
...
I1228 18:47:01.303435       1 reflector.go:240] Listing and watching *v1.Pod from k8s.io/autoscaler/cluster-autoscaler/utils/kubernetes/listers.go:174
I1228 18:47:01.303378       1 reflector.go:202] Starting reflector *v1.Node (1h0m0s) from k8s.io/autoscaler/cluster-autoscaler/utils/kubernetes/listers.go:239
I1228 18:47:01.303669       1 reflector.go:240] Listing and watching *v1.Node from k8s.io/autoscaler/cluster-autoscaler/utils/kubernetes/listers.go:239
I1228 18:47:01.801875       1 request.go:462] Throttling request took 300.987699ms, request: PUT:https://10.30.0.1:443/api/v1/namespaces/kube-system/configmaps/cluster-autoscaler-status
I1228 18:47:01.903872       1 request.go:462] Throttling request took 600.142539ms, request: GET:https://10.30.0.1:443/api/v1/nodes?resourceVersion=0
I1228 18:47:03.225966       1 leaderelection.go:199] successfully renewed lease kube-system/cluster-autoscaler
I1228 18:47:05.236322       1 leaderelection.go:199] successfully renewed lease kube-system/cluster-autoscaler
...
I1228 18:49:09.910630       1 leaderelection.go:199] successfully renewed lease kube-system/cluster-autoscaler
I1228 18:49:11.921510       1 leaderelection.go:199] successfully renewed lease kube-system/cluster-autoscaler
I1228 18:49:13.934630       1 leaderelection.go:199] successfully renewed lease kube-system/cluster-autoscaler
I1228 18:49:15.945377       1 leaderelection.go:199] successfully renewed lease kube-system/cluster-autoscaler
I1228 18:49:16.303288       1 auto_scaling.go:138] Failed to describe ASG tags for keys [k8s.io/cluster-autoscaler/enabled kubernetes.io/cluster/gnat] : RequestError: send request failed
caused by: Post https://autoscaling.us-east-2.amazonaws.com/: dial tcp: i/o timeout
F1228 18:49:16.303338       1 cloud_provider_builder.go:111] Failed to create AWS cloud provider: Failed to get ASGs: RequestError: send request failed
caused by: Post https://autoscaling.us-east-2.amazonaws.com/: dial tcp: i/o timeout
goroutine 77 [running]:
k8s.io/autoscaler/cluster-autoscaler/vendor/github.com/golang/glog.stacks(0xc420afeb00, 0xc4206180f0, 0xec, 0xed)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/vendor/github.com/golang/glog/glog.go:766 +0xa7
k8s.io/autoscaler/cluster-autoscaler/vendor/github.com/golang/glog.(*loggingT).output(0x518f4e0, 0xc400000003, 0xc4201dcbb0, 0x4e10623, 0x19, 0x6f, 0x0)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/vendor/github.com/golang/glog/glog.go:717 +0x348
k8s.io/autoscaler/cluster-autoscaler/vendor/github.com/golang/glog.(*loggingT).printf(0x518f4e0, 0x3, 0x3435f58, 0x27, 0xc420b06db0, 0x1, 0x1)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/vendor/github.com/golang/glog/glog.go:655 +0x14f
k8s.io/autoscaler/cluster-autoscaler/vendor/github.com/golang/glog.Fatalf(0x3435f58, 0x27, 0xc420b06db0, 0x1, 0x1)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/vendor/github.com/golang/glog/glog.go:1145 +0x67
k8s.io/autoscaler/cluster-autoscaler/cloudprovider/builder.CloudProviderBuilder.Build(0x7fff352eb876, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/builder/cloud_provider_builder.go:111 +0x756
k8s.io/autoscaler/cluster-autoscaler/core.NewAutoscalingContext(0xa, 0x3fe0000000000000, 0x8bb2c97000, 0x1176592e000, 0x0, 0x4e200, 0x0, 0x186a00000, 0x0, 0x7fff352eb8d3, ...)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/autoscaling_context.go:147 +0x466
k8s.io/autoscaler/cluster-autoscaler/core.NewStaticAutoscaler(0xa, 0x3fe0000000000000, 0x8bb2c97000, 0x1176592e000, 0x0, 0x4e200, 0x0, 0x186a00000, 0x0, 0x7fff352eb8d3, ...)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/static_autoscaler.go:56 +0x14d
k8s.io/autoscaler/cluster-autoscaler/core.(*AutoscalerBuilderImpl).Build(0xc42071c9c0, 0x412358, 0xc420b07850, 0x412358, 0x1a0)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/autoscaler_builder.go:71 +0x10e
k8s.io/autoscaler/cluster-autoscaler/core.NewPollingAutoscaler(0x5094380, 0xc42071c9c0, 0x78, 0x98, 0xc4203c9b80)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/polling_autoscaler.go:38 +0x35
k8s.io/autoscaler/cluster-autoscaler/core.NewAutoscaler(0xa, 0x3fe0000000000000, 0x8bb2c97000, 0x1176592e000, 0x0, 0x4e200, 0x0, 0x186a00000, 0x0, 0x7fff352eb8d3, ...)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/autoscaler.go:64 +0x5f2
main.run(0xc420af8d70)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/main.go:246 +0x263
main.main.func2(0xc42067dec0)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/main.go:344 +0x2a
created by k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/client-go/tools/leaderelection.(*LeaderElector).Run
>end of log<
```

**Comment by shiv9012** [comment]
+1

**Comment by kaazoo** [comment]
I have the same problem:
```
...
I0504 20:22:00.190888       1 main.go:288] Cluster Autoscaler 1.1.1
F0504 20:22:30.192659       1 main.go:320] Failed to get nodes from apiserver: Get https://100.64.0.1:443/api/v1/nodes: dial tcp 100.64.0.1:443: i/o timeout
goroutine 1 [running]:
k8s.io/autoscaler/cluster-autoscaler/vendor/github.com/golang/glog.stacks(0xc42068e000, 0xc42074c4e0, 0x9d, 0x19f)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/vendor/github.com/golang/glog/glog.go:766 +0xa7
k8s.io/autoscaler/cluster-autoscaler/vendor/github.com/golang/glog.(*loggingT).output(0x5618fa0, 0xc400000003, 0xc420049290, 0x52e7306, 0x7, 0x140, 0x0)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/vendor/github.com/golang/glog/glog.go:717 +0x348
k8s.io/autoscaler/cluster-autoscaler/vendor/github.com/golang/glog.(*loggingT).printf(0x5618fa0, 0x3, 0x37364df, 0x26, 0xc420ab1e88, 0x1, 0x1)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/vendor/github.com/golang/glog/glog.go:655 +0x14f
k8s.io/autoscaler/cluster-autoscaler/vendor/github.com/golang/glog.Fatalf(0x37364df, 0x26, 0xc420ab1e88, 0x1, 0x1)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/vendor/github.com/golang/glog/glog.go:1145 +0x67
main.main()
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/main.go:320 +0x4c5
```

'cluster-autoscaler' pod is restarted 6 times until it finally works:
```
kubectl get pod -n kube-system
NAME                                                                  READY     STATUS    RESTARTS   AGE
...
cluster-autoscaler-964bf7fc4-6h4ks                                    1/1       Running   6          2d
...
```

**Comment by whereisaaron** [comment]
This issue has been replicated and hasn't been fixed AFAIK.

/remove-lifecycle stale

**Comment by MaciekPytel** [comment]
ASG autodiscovery has been completely re-implemented in 1.2. Given that we still don't have an AWS mainatiner and CA 1.0 reached is no longer supported according to Kubernetes policy (https://github.com/kubernetes/community/blob/master/contributors/design-proposals/release/versioning.md#supported-releases-and-component-skew), I don't think this likely to be fixed. Would moving to CA 1.2 be an option?

**Comment by whereisaaron** [comment]
Thanks @MaciekPytel. For new clusters yes of course, but the [releases page](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler#releases) make it clear not to use 1.2 with k8s <1.10.
